src/main/scala/io/iohk/ethereum/network/rlpx/RLPxConnectionHandler.scala

```diff
@@ -136,21 +138,53 @@ class RLPxConnectionHandler(
         context.parent ! MessageReceived(message)
 
       case Failure(ex) =>
-        log.error(ex, "Cannot decode message")
+        log.error(ex, s"Cannot decode message from $peerId")
     }
 
-    def handshaked(messageCodec: MessageCodec): Receive =
+    /**
+      * Handles sending and receiving messages from the Akka TCP connection, while also handling acknowledgement of
+      * messages sent. Messages are only sent when all Ack from previous messages were received.
+      *
+      * @param messageCodec
+      * @param messagesNotSent, messages not yet sent
+      * @param messageAwaitingAck, message sent for which we are awaiting an acknowledgement by Akka TCP
+      */
+    def handshaked(messageCodec: MessageCodec,
+                   messagesNotSent: Queue[MessageSerializable] = Queue.empty,
+                   messageAwaitingAck: Option[MessageSerializable] = None): Receive =
       handleWriteFailed orElse handleConnectionClosed orElse {
         case sm: SendMessage =>
-          val out = messageCodec.encodeMessage(sm.serializable)
-          connection ! Write(out)
-          log.debug("Sent message: {}", sm.serializable)
+          if(messageAwaitingAck.isEmpty) {
+            sendMessage(messageCodec, sm.serializable)
+            context become handshaked(messageCodec, messagesNotSent, Some(sm.serializable))
+          } else
+            context become handshaked(messageCodec, messagesNotSent :+ sm.serializable, messageAwaitingAck)
```

Comment:
I agree, we should add a timeout for the `Ack`s :+1: Taking into account this losing of the `Ack` doesn't happen frequently (it doesn't seem to have happened while testing the current implementation, but i'll take a look at that while testing handling timeouts) I also think that it's not needed to resend the message.
BTW, this `Ack` is only an internal Akka message, it let's us know that the message was properly enqueued for sending by the OS (as explained in https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/io/Tcp.scala#L325), so it seems to be independent on how the other client acts.